client/Contexts/Introspect/README.md

```diff
@@ -0,0 +1,79 @@
+# Introspect
+
+This context will fetch the Strimzi admin GraphQL Schema, [perform introspection](https://graphql.org/learn/introspection/) upon it, and reduce the available capability into a UI friendly format. This state will be stored in context and available via Context consumer, but will also be accessible by hook (via `useContext`). This introspection will check the [expected](#expected-capability-schema) capabilities against what has been retrieved from Strimzi admin, resulting in [a shape which can be used across the UI](#returned-schema).
+
+## Expected capability schema
+
+A GraphQL schema is made up of `queries`, `types`, `mutations` and `subscriptions`. Together they describe what the server hosts and is capable of, but introspection is required to understand the relationship between these constructs, and what `mutation` for example will update a particular `type`.
+
+To drive this introspection process, the Strimzi ui defines what it requires/expects to operate at an entity level. This approach will define what GraphQL `types` are expected, the fields those types should contain (and what types they are), the operations (`mutations`) that can be performed on a given `type`, and what `subscriptions` are available for a given type as well. Using the example of a Topic, this would look as follows:
+
+```
+{
+    Topic: {
+        type: 'Topic', // the GraphQL type name
+        fields: {
+            name: GQL_STRING, // the expected GraphQL type - could be string or object - which is checked if is available and thus query-able
```

Comment:
If we have an "unreliable" schema (hence the introspection) - would it be simpler to have a translation layer?

As in - components call `buildQuery` with the "complete" query, and fields are filtered back out? And the same way with a `parseResults`, it results in an object with `null`s for bits that are missing?

My concern is that the components now have to build composed queries based off fields/capabilities.

What would be great is if introspection told the component what "capabilities" can be used (and those control features) - but the data in/out was just filtered out so that you just handled null if something you needed was missing?

Reply:
> What would be great is if introspection told the component what "capabilities" can be used (and those control features)

That is the intent here - this model descirbes what is expected, which can then be checked with flags being enabled /disabled for example. The only usage so far is to show/hide whole pages, but this could be used at a feature flag level as well.

I must admit, I have not considered a partial schema case (and returning nulls/empty values) for a query. Let me look into this and see what would need to change to support it.